general/testing.md

```diff
@@ -1,5 +1,19 @@
 # Recommendations for testing
 
+## Testing UI components
+
+-   Snapshots SHOULD only be used as "smoke tests": they are very good for verifying that a component renders at all, but they are _not_ good for communicating intent or what is important. For example, when something in a large snapshot changes, there is nothing intrinsic in the snapshot to communicate whether any particular part of it is important.
+-   Shipping test attributes to the client is overtly bad because it increases network payloads, parse times, DOM bloat etc, and ends up becoming an unofficial API.
+    -   Therefore, we MUST NOT send "data-testid" to the client.
+-   The code and markup that is exercised in the test environment should match that which is used in the production environment, otherwise you are not actually testing the thing that you would like to test.
```

Comment:
This was something that happened in the past, but I don't have a clear feeling of wether it can be changed or not... if it could be changed, how important do you think this is, shipping and testing the same exact code?

After all, we're doing all sorts of transformations already...

Reply:
I'd say:

    don't ship test junk to client > ship exactly the same bytes to client

ie. desirable, but not essential, which is why it appears further down the list.

This is going to be subtle, but another thing about "data-testid" though that makes it seem like a potential code smell to me is this:

A "data-testid" says "this thing is important to me". But if it is important, why isn't it targetable in some user-visible way? That is, users care about behavior, content, and behavior of specific things (buttons, sliders, charts etc). If the only way I can label something as important is via *a mechanism that is invisible to the user* (not even in the source code on the client), then there is a risk that I might be testing an implementation detail and not something which a user actually cares about. And if QA folks start leveraging these "data-testid" attributes as a defacto API, then we wind up in a situation where QA may be asserting things at the layer of implementation details, and which isn't the layer that users interact with and care about.

Or maybe another way to say that is that ideally these should all be aligned:

    what users care about === what you actually test === the API you use to test

with the obvious observation that the definition of "API" can include markup etc, not just programming interfaces.

Reply:
Sounds good to me, just trying to present alternate points for discussion :)

I think it would be good to document this somewhere as rationale. This reasoning carries more weight than "historical reasons prevent us from..."